### PR TITLE
[XLA] Add half types to the new batchnorm op

### PR DESCRIPTION
**Pull request by DavidNorman**
The new Batchnorm operations should allow a backend to operate at half precision as well as full float32 precision.

### CODE REVIEW

**Review by sanjoy — CHANGES_REQUESTED** [review]
I'm unsure if this is in general a good idea -- does `U` affect the precision of the intermediate computation does by batch norm?  If yes, using a narrow type increases the chances of overflow.

**Comment by DavidNorman** [comment]
a particular backend does not need to support any specific value in any of the fields in its kernal registration.

I take it you are referring to the values of the reserved_fields, which are presumably a hack to get intermediate values from the fwd to the bwd pass.

perhaps it would be better if my change modified the type of the mean/scale/variance/offset parameters to 'T', which is really what I was trying to achieve?

**Comment by sanjoy** [comment]
> perhaps it would be better if my change modified the type of the mean/scale/variance/offset parameters to 'T', which is really what I was trying to achieve?

Or you could create a separate attribute `S` that allows narrow types and have the `mean`/`scale`/`variance`/`offset` parameters be of this type.  I don't know if there is way to make this backward compatible though -- we'd want `S` to default to the value of `T`, whatever that is.

Btw, OOC, why can't you have these parameters be `float`?  Can't you cast them back to `half` / `bfloat` in the implementation?

**Comment by sanjoy** [comment]
Btw, Reed is probably the best person to review this.

**Comment by reedwm** [comment]
Is there a purpose of allowing `mean`/`scale`/`variance`/`offset` to be float16 or bfloat16? Nvidia observed these values should be float32 for convergence, and there seems to be no upside to having them be a lower precision.

**Comment by DavidNorman** [comment]
we have observed that things converge fine with stochastic rounded float16.

**Comment by DavidNorman** [comment]
i don't understand why you would limit the definition of the general operation, when the registration of the device specific kernel operation can place per device type limitations.

**Comment by DavidNorman** [comment]
regarding 'Btw, OOC, why can't you have these parameters be float? Can't you cast them back to half / bfloat in the implementation?', surely the logical conclusion of this is that all floating point numbers  should be specified as double precision.

**Comment by reedwm** [comment]
/CC @alextp, what do you think of this change? Is there precedent to allowing a type in the OpDef, but not implementing the change in any kernel? I would recommend to users to always keep `mean`/`scale`/`variance`/`offset` in float32, but users may decide they would rather have them in float16.

I think this will also make the error messages worse if a user does pass a float16 value to `mean`/`scale`/`variance`/`offset`.

**Comment by DavidNorman** [comment]
Incidentally, the reason I did this was because the old FusedBatchNormV2 was recently replaced by BatchNormV3.  

BatchNormV2 has the following signature:

```

**Comment by alextp** [comment]
@reedwm I don't think it should be in the opdef's business to specify the dtypes too narrowly. It's best to leave that to kernel registrations. Many ops technically support dtypes for which we don't have kernels on some devices (say, Add which supports string dtype but not on the gpu)

File: tensorflow/core/ops/nn_ops.cc [hunk 1]
```diff
@@ -214,7 +214,7 @@ REGISTER_OP("FusedBatchNormV3")
     .Output("reserve_space_2: U")
     .Output("reserve_space_3: U")
     .Attr("T: {half, bfloat16, float}")
-    .Attr("U: {float}")
+    .Attr("U: {half, float}")
```
Comment (reedwm) [inline tensorflow/core/ops/nn_ops.cc]: Why not support bfloat16 here, when it's supported in FusedBatchNormV2?

**Review by reedwm — CHANGES_REQUESTED** [review]

**Comment by reedwm** [comment]
> I don't think it should be in the opdef's business to specify the dtypes too narrowly

In that case, this changes looks good. I just have one comment.

File: tensorflow/core/ops/nn_ops.cc (hunk 1 above)
Comment (DavidNorman) [reply tensorflow/core/ops/nn_ops.cc]: i totally agree.  mistake.  will correct.

**Comment by DavidNorman** [comment]
thanks guys.  will correct the bfloat thing.

**Comment by DavidNorman** [comment]
i hope all the tests pass.  we don't run the GPU tests in our CI suite, so I can't tell ahead of time if they will break.  I'm guessing not, given that the public python part of the batchnorm interface will still create float32 stats variables.

**Comment by DavidNorman** [comment]
I would be keen to hear about the long term plan with regard to this commit.  We have the same issue, in the sense that our device-side library uses inv-sqrt as its stats outputs (apparently it is something to do with dynamic ranges).

```
commit 4b2be43f63b3b11e6d8d5a6648330e42bf5600cb
Author: Sanjoy Das <sanjoy@google.com>
Date:   Fri Feb 1 15:39:18 2019 -0800

    Fix TF/XLA lowering for FusedBatchnorm training on GPU
    
    The FusedBatchnorm TensorFlow operation has two outputs, reserve_space_{1|2},
    that contain unspecified results; in practice on CPUs the they contain the mean
    and the variance and on GPUs they contain the mean the *inverse sqrt of the
    variance* (what cudnn returns).  This is unfortunate for XLA because without a
    spec describing what these outputs are, we can't soundly replace these outputs
    with XLA computed values.
    
    For now we rely on the in-practice values returned in reserve_space_{1|2} and
    add a test case checking that these values are indeed correct.  In the future we
    may consider some of the cleaner fixes outlined in the tracking bug.
    
    PiperOrigin-RevId: 232050515
```

**Review by reedwm — APPROVED** [review]

**Comment by sanjoy** [comment]
> I would be keen to hear about the long term plan with regard to this commit. We have the same issue, in the sense that our device-side library uses inv-sqrt as its stats outputs (apparently it is something to do with dynamic ranges).

Nobody is working on this at the moment, but there are the two other ideas we discussed (from the internal bug report):

@timshen91 had this interesting idea: we could change the contract of `reserve_space_{1|2}` to be that it either contains the cudnn implementation defined values or a sentinel value (e.g a scalar 0).  Then in the cudnn implementations of `FusedBatchNorm{Grad}` training we could check the shape of the `reserve_space_{1|2}` and act accordingly.

We could even go one step further and have `FusedBatchNorm{Grad}` consume/return an additional host side integer that tells us how to interpret `reserve_space_{1|2}`.  The cudnn kernel for `FusedBatchNormGrad` would then check if this integer was, say, `kCudnnBuffer` and only then pass it to the cudnn backwards batch norm.

**Comment by DavidNorman** [comment]
I will remove the extra limitation on the format, because it is failing the Ubuntu CC tests.

**Comment by DavidNorman** [comment]
In retrospect, i have not changed the data format line at all.  Is this is an error in the Ubuntu CC test where it is comparing an earlier registration to the current code base?

ps.  I think I have addressed all the changes now.

**Comment by reedwm** [comment]
This is a bizarre error. It looks like ops_history.v1.pbtxt incorrectly does not currently list the allowed_values for the data_format attribute in FusedBatchNormGradV2. Not sure how this happened.

I'll try to fix ops_history.v1.pbtxt. Sorry for the delay.

**Comment by reedwm** [comment]
Try merging past 4c7e2edfea75200d2e4c20e32c73a8a7fb7f764b and running the tests again.

**Comment by DavidNorman** [comment]
cheers - doing that now

**Comment by DavidNorman** [comment]
done.

**Review by reedwm — DISMISSED** [review]

**Comment by reedwm** [comment]
Ugh, now the compatibility test is complaining that for the FusedBatchNormGradV2 and FusedBatchNormGradV3 ops, the type of "scale" is changed from float to U. I don't think this actually breaks backwards compatibility, as previously, U could only be float anyway, so any old GraphDefs will not be broken.

/CC @alextp, what's the best approach here? Creating a new FusedBatchNormGradV4 op? (This is the best approach IMO). Force submitting? Trying to fix the test? (Although it looks like fixing the test is tricky and I currently don't have time).

**Comment by alextp** [comment]
@reedwm the solution is not reuse U, and instead make a new dtype parameter for the things which used to be float. You can still register kernels which only support U=new_parameter.

**Comment by reedwm** [comment]
@alextp the problem with adding a new dtype parameter is that then, `scale` could be a different dtype from the rest of the parameters like `reserve_space_1`, which is not allowed.

For example, currently, the op looks like this (I removed some unnecessary lines):

```
REGISTER_OP("FusedBatchNormGradV3")
    .Input("scale: float")
    .Input("reserve_space_1: U")
    .Attr("U: {float}")
```

This PR changes the op to look like this:

```
REGISTER_OP("FusedBatchNormGradV3")
    .Input("scale: U")
    .Input("reserve_space_1: U")
    .Attr("U: {half, bfloat16, float}")
```

In both cases, `scale` and `reserve_space` have the same dtype, which is a requirement. But If we introduce a new dtype parameter, `scale` and `reserve_space` could be different, which is undesired:

```
REGISTER_OP("FusedBatchNormGradV3")
    .Input("scale: V")
    .Input("reserve_space_1: U")
    .Attr("V: {half, bfloat16, float}")
    .Attr("U: {half, bfloat16, float}")
```

**Comment by alextp** [comment]
@reedwm I understand the issue. This is why I encourage you to relax the opdef restriction (by adding another attribute) and instead enforce this property by not registering kernels for the invalid combinations.

Doing so preserves backward compatibility, in that no existing graph becomes invalid, which is the property that the backwards compatibility test tries to enforce but fails here.

**Comment by reedwm** [comment]
@alextp thank you for the clarifcation.

@DavidNorman, can you update `FusedBatchNormGradV2` and `FusedBatchNormGradV3` to add a new attribute for `scale`, as @alextp suggested?

**Comment by DavidNorman** [comment]
sure thing.  that makes perfect sense.

**Comment by DavidNorman** [comment]
done :)

**Review by reedwm — DISMISSED** [review]

**Comment by rthadur** [comment]
@DavidNorman can you please check this link https://source.cloud.google.com/results/invocations/75f4a655-9bd0-4115-baf5-1bfa96838209/log
for logs , where ubuntu CC builds are failing.

**Comment by DavidNorman** [comment]
i see.  it needs a default type.  i'll add that right away

**Comment by DavidNorman** [comment]
done

**Comment by DavidNorman** [comment]
@rthadur done

**Review by reedwm — DISMISSED** [review]

**Comment by DavidNorman** [comment]
ok - i can see that i'm going to have to look into the data format restriction - make it back into no restriction.  and also fix up this default value thing.

**Comment by DavidNorman** [comment]
@reedwm  @rthadur 

I've added the type constraint to the kernels for CPU and GPU in the hope that the grappler GPU test will be ok with that.  The other failures are the return of the data_format issue which  isn't something that has been touched by this pull request.

**Review by reedwm — APPROVED** [review]

**Comment by rthadur** [comment]
@DavidNorman thank you, there are few more tests failing for ubuntu CC here : https://source.cloud.google.com/results/invocations/2499aa6a-de8d-4f0c-b2a1-e3a1a03f2b91/log
can you please check

**Comment by DavidNorman** [comment]
@rthadur this failing test is testing something that I have not changed.  It is complaining that the batchnorm data format is more constrained than for the current head.  however, this is not true.  the data format has not changed.  i think that the test needs to be checked.  perhaps it is testing against the release version of the framework, not the master version.  i do not think that I should be removing the constraint that someone else has added just to pass the test.

**Comment by rthadur** [comment]
@DavidNorman will try to pull this in.

**Comment by rthadur** [comment]
@DavidNorman this is failing internally with same errors , @reedwm any thoughts how we proceed with this ?

**Comment by reedwm** [comment]
Whoops, I thought I fixed this in 4c7e2edfea75200d2e4c20e32c73a8a7fb7f764b, but I only fixed it for the gradient ops. Didn't realize it affected the non-gradient ops.

I will fix the bug preventing the tests here from working properly.

**Comment by rthadur** [comment]
@reedwm  thank you , will keep merging this on hold.

**Comment by reedwm** [comment]
5c7b8ea9b23a60efbbe20d3bb3b679a6be381924 should fix this.

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by AyanmoI** [comment]
Correct me if I am wrong, but from what i understand, this change would allow the stats variables to be `half` of `bfloat` in addition to `float`. In that case what is the canonical form in the bridge? The GPU backend expects these variables to in fp32 for mixed precision computation. Do all backends now need to add passes to cast the stat variables according to their requirements? If not, then is the canonical form in the bridge still fp32? In that case the bridge needs to handle non-float types by inserting coverts which the backend can eventually choose to eliminate. Sorry if this has already been discussed earlier and  I missed it.

**Comment by reedwm** [comment]
I am unfamiliar with the tf2xla bridge details. /CC @sanjoy.

Note even with this change, a BatchNormalization layer will still keep stats variables in float32, and nearly all models would still have the stats variables in float32.

**Comment by DavidNorman** [comment]
@AyanmoI This change redefines the acceptable types of the stats variables.  It does not change the types which are actually registered by each backend, so the GPU registration remains fp32 only, and the GPU will fail if the user attempts to provide an fp16 stat.   This was true before, just the failure was because of a mismatched op registration, not kernel registration.

The tf2xla bridge can do what it likes.  As it happens, it does the right thing with non-fp32 types,  but if it was fixed at fp32 for some mad reason, then we could override the tf2xla implementation with a custom one for our backend.

**Comment by AyanmoI** [comment]
> @AyanmoI This change redefines the acceptable types of the stats variables. It does not change the types which are actually registered by each backend, so the GPU registration remains fp32 only, and the GPU will fail if the user attempts to provide an fp16 stat. This was true before, just the failure was because of a mismatched op registration, not kernel registration.
I agree. However, IMO with this change, we don't even need to error out. All we need to do is decide on a canonical representation of the batchnorm op at tf2xla level. If we leave it as it is, then backends like GPU can insert converts through a pass if it sees the stats variables in (say) fp16. Or, we can have converts to fp32 in tf2xla and backends which want the stats variables in `Half` can insert another convert in their backend and `algebraic_simplifier` would eliminate the convert pair eventually.
> The tf2xla bridge can do what it likes. As it happens, it does the right thing with non-fp32 types, but if it was fixed at fp32 for some mad reason, then we could override the tf2xla implementation with a custom one for our backend.

This PR in itself is fine but I believe we can make changes to the bridge or backends such that no errors occur. We shouldn't throw an error anyway since {fp16,bfloat}->fp32 is an upcast and it should just work from the users perspective.

**Comment by DavidNorman** [comment]
We've decided to abandon this and do something clever lower down.  there are too many other things higher up in the python code which would need to be modified too.
